c++11/include/opentracing/tracer.h

```diff
@@ -0,0 +1,192 @@
+// the start timestamp, specify Span References, and make a single Tag or
+// multiple Tags available at Span start time.
+//
+// StartSpan() callers should look at the StartSpanOption interface and
+// implementations available in this library.
+struct StartSpanOptions {
+  SystemTime start_system_timestamp;
+  SteadyTime start_steady_timestamp;
+  std::vector<std::pair<SpanReferenceType, const SpanContext*>> references;
+  std::vector<std::pair<std::string, Value>> tags;
+};
+
+// StartSpanOption instances (zero or more) may be passed to Tracer.StartSpan.
+class StartSpanOption {
+ public:
+  StartSpanOption(const StartSpanOption&) = delete;
+
+  virtual ~StartSpanOption() = default;
+
+  virtual void Apply(StartSpanOptions& options) const = 0;
+
+ protected:
+  StartSpanOption() = default;
+};
+
+// Tracer is a simple, thin interface for Span creation and SpanContext
+// propagation.
+class Tracer {
+ public:
+  virtual ~Tracer() = default;
+
+  // Create, start, and return a new Span with the given `operationName` and
+  // incorporate the given StartSpanOption `option_list`.
+  //
+  // A Span with no SpanReference options (e.g., opentracing::ChildOf() or
+  // opentracing::FollowsFrom()) becomes the root of its own trace.
+  //
+  // Examples:
+  //
+  //     opentracing::Tracer& tracer = ...
+  //
+  //     // The root-span case:
+  //     auto span = tracer.StartSpan("GetFeed")
+  //
+  //     // The vanilla child span case:
+  //     auto span = tracer.StartSpan(
+  //         "GetFeed",
+  //         {opentracing::ChildOf(parentSpan.context())})
+  //
+  //     // All the bells and whistles:
+  //     auto span = tracer.StartSpan(
+  //         "GetFeed",
+  //         {opentracing::ChildOf(parentSpan.context()),
+  //         opentracing::Tag{"user_agent", loggedReq.UserAgent},
+  //         opentracing::StartTimestamp(loggedReq.timestamp())})
+  //
+  std::unique_ptr<Span> StartSpan(
+      const std::string& operation_name,
+      std::initializer_list<option_wrapper<StartSpanOption>> option_list = {})
+      const {
+    StartSpanOptions options;
+    options.start_system_timestamp = SystemClock::now();
+    options.start_steady_timestamp = SteadyClock::now();
+    for (const auto& option : option_list) option.get().Apply(options);
+    return StartSpanWithOptions(operation_name, options);
+  }
+
+  virtual std::unique_ptr<Span> StartSpanWithOptions(
+      const std::string& operation_name,
+      const StartSpanOptions& options) const = 0;
+
+  // Inject() takes the `sc` SpanContext instance and injects it for
+  // propagation within `carrier`. The actual type of `carrier` depends on
+  // the value of `format`.
+  //
+  // OpenTracing defines a common set of `format` values (see BuiltinFormat),
+  // and each has an expected carrier type.
+  //
+  // Returns true on success.
+  virtual bool Inject(const SpanContext& sc, CarrierFormat format,
+                      const CarrierWriter& writer) = 0;
```

Comment:
Should `Inject` have a more descriptive return code? Perhaps something similar to [this](https://github.com/opentracing/opentracing-cpp/blob/master/opentracing/tracer.h#L145)?